### PR TITLE
Feat/remove fetcher for uploads 343

### PR DESCRIPTION
**Pull request by microsoftly**
addresses #343 

The fossa api exposes/will expose an endpoint `/api/cli/supported_versions` that returns an an array of `vX` versions that are currently supported. We can cross reference that list with what the CLI currently knows it can support and select the best version to support (as of now, that selection criterion is simply what is newest). In theory, this should allow us to merge at our leisure 

I'm forgoing programatic testing of selecting which API is selected and the upload payload until we get a stable mock server implementation.

### CODE REVIEW

File: cmd/fossa/cmd/analyze/analyze.go [hunk 1]
```diff
@@ -133,6 +133,68 @@ func Do(modules []module.Module) (analyzed []module.Module, err error) {
...
 
 func uploadAnalysis(normalized []fossa.SourceUnit) error {
+	supportedApiVersions, err := fossa.GetSupportedApiVersions()
+	if err != nil {
+		return err
+	}
+
+	// lists out all compatible versions, where versions listed earlier in the list should be preferred to be used over later appearing versions
+	compatibleVersionsByPreference := []string{"v1", "v0"}
+	for _, compatibleVersion := range compatibleVersionsByPreference {
```
Comment (liftM) [inline cmd/fossa/cmd/analyze/analyze.go]: This code is pretty crazy, and we don't need it if we change the interface to `fossa.GetLatestSupportedAPIVersion()`.

File: cmd/fossa/cmd/analyze/analyze.go [hunk 2]
```diff
@@ -133,6 +133,68 @@ func Do(modules []module.Module) (analyzed []module.Module, err error) {
...
+	uploadBody.Project.Url = config.ProjectURL()
+	uploadBody.Project.Title = config.Title()
+	uploadBody.Project.Jira = config.JIRAProjectKey()
+
+	uploadBody.Revision.Link = config.Link()
+	uploadBody.Revision.Id = config.Revision()
+
+	uploadBody.Vcs.Reference = config.Branch()
+
+	if config.GitVcsExists() {
```
Comment (liftM) [inline cmd/fossa/cmd/analyze/analyze.go]: This is a good place for `config.TryStrings`.

File: api/fossa/uploadv1.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	endpoint, err := url.Parse("/api/cli/v1/build")
+	if err != nil {
+		return Locator{}, errors.New("Failed to generate upload URL")
+	}
+	log.WithField("endpoint", endpoint.String()).Debug("uploading build")
+
+	res, statusCode, err := Post(endpoint.String(), payload)
+	log.WithField("response", res).Debug("build upload completed")
+
+	if statusCode == 428 {
```
Comment (liftM) [inline api/fossa/uploadv1.go]: Is this behaviour still valid in the `v1` API route?

File: api/fossa/uploadv1.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	log.Debug("build upload succeeded")
+
+	var unmarshalled struct {
+		Locator string
+		Error   string
+	}
+	err = json.Unmarshal([]byte(res), &unmarshalled)
+	if err != nil || unmarshalled.Error != "" {
+		return Locator{}, errors.Errorf("Error response from API: %s", res)
```
Comment (zlav) [inline api/fossa/uploadv1.go]: We should probably return the error here as well. The error response here is very similar to line 89 without including the error, although I would be interested to know if you have a reason for this.

File: cmd/fossa/cmd/analyze/analyze.go (hunk 1 above)
Comment (zlav) [reply cmd/fossa/cmd/analyze/analyze.go]: I agree. It looks like we are looping through `compatibleVersionsByPreference` and then later switching on the same values. Is there a reason to check for a supported version twice?

File: cmd/fossa/cmd/analyze/analyze.go [hunk 5]
```diff
@@ -133,6 +133,68 @@ func Do(modules []module.Module) (analyzed []module.Module, err error) {
...
+		uploadBody.Vcs.Type = "git"
+	} else {
+		uploadBody.Vcs.Type = "none"
+	}
+
+	locator, err := fossa.UploadV1(uploadBody)
+	display.ClearProgress()
+	if err != nil {
+		log.Fatalf("Error during upload: %s", err.Error())
+		return err
```
Comment (zlav) [inline cmd/fossa/cmd/analyze/analyze.go]: This code is unreachable if the line above it logs fatal. I'm not a fan of logging fatal and would prefer to return the error here instead, same with line 216 where this was done before.

File: api/fossa/uploadv1.go (hunk 3 above)
Comment (microsoftly) [reply api/fossa/uploadv1.go]: Yes, as of now the goal of the v1 API route was to provide an equivalent output for the a different input. The goal was to keep the 428 response.

File: cmd/fossa/cmd/analyze/analyze.go (hunk 1 above)
Comment (microsoftly) [reply cmd/fossa/cmd/analyze/analyze.go]: I'd be okay with moving this to be scoped within the fossa API package. 

My goal in having this loop was to handle situations where users have an older version of `cli`   or on prem users that may have an older version of core.

@zlav Its not checking supported version twice, it's determining which versions that core supports are also supported by this build of `cli`

File: api/fossa/apiversion.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+func GetSupportedApiVersions() ([]string, error) {
+	endpoint, err := url.Parse("/api/cli/supported_versions")
+	if err != nil {
+		return []string{}, err
+	}
+
+	res, statusCode, err := Get(endpoint.String())
+	if err != nil {
+		return []string{}, errors.Wrap(err, "Could not get supported API versions.")
```
[the remark this answers is not in the document]
Comment (microsoftly) [reply api/fossa/apiversion.go]: I've verified that `404` response results in `err == nil`

File: cmd/fossa/cmd/analyze/analyze.go (hunk 2 above)
Comment (microsoftly) [reply cmd/fossa/cmd/analyze/analyze.go]: I don't see how that would help here. This is a clear `if` `else` case for setting a value where there is no default or initial value to override

File: api/fossa/upload.go [hunk 7]
```diff
@@ -1,114 +1,94 @@
...
-		q.Add("jiraProjectKey", options.JIRAProjectKey)
-	}
-	if options.Link != "" {
-		q.Add("link", options.Link)
-	}
-	if options.Team != "" {
-		q.Add("team", options.Team)
-	}
+func uploadAnalysisV1(normalized []SourceUnit) (Locator, error) {
+	display.InProgress("Uploading analysis...")
```
Comment (liftM) [inline api/fossa/upload.go]: The `display.InProgress` calls should go in the `cmd` packages, not the `api`.

File: api/fossa/upload.go [hunk 8]
```diff
@@ -1,114 +1,94 @@
...
 	}
-	log.WithField("endpoint", endpoint.String()).Debug("uploading build")
 
-	res, statusCode, err := Post(endpoint.String(), payload)
-	log.WithField("response", res).Debug("build upload completed")
+	uploadBody.Project.Id = config.Project()
+	uploadBody.Project.Team = config.Team()
+	uploadBody.Project.Url = config.ProjectURL()
+	uploadBody.Project.Title = config.Title()
+	uploadBody.Project.Jira = config.JIRAProjectKey()
```
Comment (liftM) [inline api/fossa/upload.go]: [nit] `s/Jira/JIRA`

File: api/fossa/upload.go [hunk 9]
```diff
@@ -1,114 +1,94 @@
...
-	if err != nil {
-		return Locator{}, errors.New("Failed to generate upload URL")
+	uploadBody := V1UploadBody{
+		Analysis: normalized,
 	}
-	log.WithField("endpoint", endpoint.String()).Debug("uploading build")
 
-	res, statusCode, err := Post(endpoint.String(), payload)
-	log.WithField("response", res).Debug("build upload completed")
+	uploadBody.Project.Id = config.Project()
```
Comment (liftM) [inline api/fossa/upload.go]: [nit] `s/Id/ID`

File: api/fossa/upload.go [hunk 10]
```diff
@@ -1,114 +1,94 @@
...
-		// TODO: handle "Managed Project" workflow
-		return Locator{}, errors.New("invalid project or revision; make sure this version is published and FOSSA has access to your repo (to submit a custom project, set Fetcher to `custom` in `.fossa.yml`)")
-	} else if statusCode == 403 {
-		return Locator{}, errors.New("you do not have permission to upload builds for this project")
-	} else if err != nil {
-		return Locator{}, errors.Wrap(err, "could not upload")
-	}
-	log.Debug("build upload succeeded")
+	uploadBody.Revision.Link = config.Link()
+	uploadBody.Revision.Id = config.Revision()
```
Comment (liftM) [inline api/fossa/upload.go]: [nit] `s/Id/ID`

File: api/fossa/uploadv1.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+type V1UploadBody struct {
+	VCS struct {
+		Type      string
+		Reference string
+	}
+	Project struct {
+		Title string
+		Id    string
```
Comment (liftM) [inline api/fossa/uploadv1.go]: [nit] `s/Id/ID`

File: api/fossa/uploadv1.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+		Url   string
```
Comment (liftM) [inline api/fossa/uploadv1.go]: [nit] `s/Url/URL`

File: api/fossa/uploadv1.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+		Jira  string
```
Comment (liftM) [inline api/fossa/uploadv1.go]: [nit] `s/Jira/JIRA`

File: api/fossa/uploadv1.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+		Team  string
+	}
+	Revision struct {
+		Id   string
```
Comment (liftM) [inline api/fossa/uploadv1.go]: [nit] `s/Id/ID`

File: api/fossa/uploadv1.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+		Link string
+	}
+	Meta struct {
+		v string
```
Comment (liftM) [inline api/fossa/uploadv1.go]: This should be `V`, otherwise `json.Marshal` will not pick up an unexported field.

**Pull request merged**